src/Networking/real_world_interface.cpp

```diff
@@ -1,81 +1,160 @@
+	// This is a bit on the conservative side, but we heard an ominous popping sound at 20000.
+	int16_t max_pwm = 15000;
+	if (abs(right_pwm) > max_pwm) {
+		std::cout << "WARNING: requested too-large right PWM " << right_pwm << std::endl;
+		right_pwm = max_pwm * (right_pwm < 0 ? -1 : 1);
+	}
+	if (abs(left_pwm) > max_pwm) {
+		std::cout << "WARNING: requested too-large left PWM " << left_pwm << std::endl;
+		left_pwm = max_pwm * (left_pwm < 0 ? -1 : 1);
+	}
+
+	CANPacket p;
+	uint8_t motor_group = 0x04;
+	AssemblePWMDirSetPacket(&p, motor_group, DEVICE_SERIAL_MOTOR_CHASSIS_FL, left_pwm);
+	sendCANPacket(p);
+	AssemblePWMDirSetPacket(&p, motor_group, DEVICE_SERIAL_MOTOR_CHASSIS_FR, right_pwm);
+	sendCANPacket(p);
+	AssemblePWMDirSetPacket(&p, motor_group, DEVICE_SERIAL_MOTOR_CHASSIS_BL, left_pwm);
+	sendCANPacket(p);
+	AssemblePWMDirSetPacket(&p, motor_group, DEVICE_SERIAL_MOTOR_CHASSIS_BR, right_pwm);
+	sendCANPacket(p);
+	return true;
 }
 
 // Everything below this line still needs to be implemented
-
+points_t last_landmarks = zero_landmarks;
+uint32_t last_frame_no = 0;
+cv::Mat frame;
 points_t readLandmarks() {
-  points_t lms({});
-  for (int i = 0; i < 11; i++) {
-    point_t lm({0.0,0.0,0.0});
-    lms.push_back(lm);
-  }
-  return lms;
+	// camera might not be open if opening camera failed in init_world_interface.
+	// if it is not, return a list of 11 zero points.
+	if (ar_cam.isOpen()) {
+		// if we have a next frame, retrieve it and perform AR tag detection.
+		if(ar_cam.hasNext(last_frame_no)){
+			ar_cam.next(frame, last_frame_no);
+			std::vector<AR::Tag> tags = ar_detector.detectTags(frame);
+			log(LOG_DEBUG, "readLandmarks(): %d tags spotted\n", tags.size());
+
+			// build up map with first tag of each ID spotted.
+		    points_t output(NUM_LANDMARKS);
+			std::map<int, cv::Vec3d> first_tag;
+			for(AR::Tag tag : tags){
+				int id = tag.getMarker().getId();
+				if(first_tag.find(id) == first_tag.end()){
+					first_tag[id] = tag.getCoordinates();
+				}
+			}
+
+			// for every possible landmark ID, go through and if the map contains a value for
+			// that ID, add it to the output array (doing appropriate coordinate space
+			// transforms). If not, add a zero point.
+			for(int i = 0; i < NUM_LANDMARKS; i++){
+				if(first_tag.find(i) != first_tag.end()){
+					cv::Vec3d coords = first_tag[i];
+					// if we have extrinsic parameters, multiply coordinates by them to do
+					// appropriate transformation. If not, just change the coordinate axes (no
+					// rotation/translation)
+					if (ar_cam.hasExtrinsicParams()) {
+						cv::Vec4d coords_homogeneous = {coords[0], coords[1], coords[2], 1};
+						cv::Mat transformed =
+							ar_cam.getExtrinsicParams() * cv::Mat(coords_homogeneous);
+						output[i] = {transformed.at<double>(0, 0),
+									 transformed.at<double>(1, 0),
+									 1.0};
+					} else {
+						// just account for coordinate axis change.
```

Comment:
It might be good to expand on this comment a bit. Just to say what the camera coordinate axes are vs. the rover frame coordinate axes.